MariaDB/server
# MDEV-35049: Improve adaptive hash index scalability

- [x] *The Jira issue number for this PR is: MDEV-35049, MDEV-35312*
## Description
This also includes #3574 and #3627 until they have been merged to 10.6.

I suggest to read the individual [commit mesages|https://github.com/MariaDB/server/pull/3562/commits] (`git log origin/10.6..origin/10.6-MDEV-35190`).

This is a rewrite of the InnoDB adaptive hash index, as well as a rewrite of record comparisons so that heap memory allocation as well as calls to `rec_get_offsets()` will be avoided.

`MEM_HEAP_BTR_SEARCH` and `mem_block_info_t::free_block` are removed. We implement a custom memory allocator for the adaptive hash index entries (`ahi_node`, renamed from `ha_node_t`). The `mem_heap` subsystem is simplfiied.

We also make `rec_fold()` compute CRC-32C on a contiguous range of bytes, from the start of the record to the last included byte of the last field. In `dtuple_fold()` we compute the checksum field by field.

The function `page_cur_search_with_match_bytes()`, which can benefit `INSERT` into the end of an index page, will be enabled even when `innodb_adaptive_hash_index=OFF`.

`btr_search_guess_on_hash()` will return `BTR_CUR_HASH_FAIL` in fewer cases. This will lead to less frequent rebuild of the adaptive hash index and therefore better performance in a read-only workload.

Many fields related to comparing records and the adaptive hash index are consolidated into a single `n_bytes_fields` or `left_bytes_fields`. A redundant set of AHI related parameters is removed from `buf_block_t`. The `buf_block_t::ahi_left_bytes_fields` will use `Atomic_relaxed<uint32_t>`.
## Release Notes
The performance of `innodb_adaptive_hash_index=ON` was improved at high concurrency.
## How can this PR be tested?
```sh
./mtr --parallel=auto --big-test --force --mysqld=--loose-innodb-adaptive-hash-index=ON --max-test-fail=0
```
I tested this with and without `cmake -DWITH_INNODB_AHI=OFF` as well as with `cmake -DWITH_INNODB_ROOT_GUESS=OFF`.

With the `innodb-hashtest.sh` that @montywi attached to [MDEV-35049](https://jira.mariadb.org/browse/MDEV-35049) I no longer see any waits for the `btr_sea::partition::latch`, and there are no calls to `btr_sea::partition::insert` during the workload. This means that the adaptive hash index remains constant during a constant read-only workload where the data fits in the buffer pool.

This needs to be tested with a larger workload that originally reproduced the scalability issue.
## Basing the PR against the correct MariaDB version

Part of this could be applicable to 10.5, but the patch is currently is based on the 10.6 branch, because it depends on many performance improvements that are only available starting with 10.6.
## PR quality check
